diff -up nfs-utils-2.3.3/utils/mount/stropts.c.orig nfs-utils-2.3.3/utils/mount/stropts.c
--- nfs-utils-2.3.3/utils/mount/stropts.c.orig	2022-02-14 11:28:51.570084952 -0500
+++ nfs-utils-2.3.3/utils/mount/stropts.c	2022-02-14 11:29:16.174450628 -0500
@@ -966,7 +966,9 @@ fall_back:
 	if ((result = nfs_try_mount_v3v2(mi, FALSE)))
 		return result;
 
-	errno = olderrno;
+	if (errno != EBUSY && errno != EACCES)
+		errno = olderrno;
+
 	return result;
 }
 
